pkg/scheduler/plugins/gpushare/device_cache.go

```diff
@@ -0,0 +1,132 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gpushare
+
+import (
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	quotav1 "k8s.io/apiserver/pkg/quota/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	apiext "github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+)
+
+type nodeDeviceCache struct {
+	lock sync.Mutex
+	// nodeDeviceInfos stores nodeDevice for each node
+	// and uses node name as map key.
+	nodeDeviceInfos map[string]*nodeDevice
+}
+
+type nodeDevice struct {
+	lock        sync.Mutex
+	deviceTotal map[schedulingv1alpha1.DeviceType]*deviceResources
+	deviceFree  map[schedulingv1alpha1.DeviceType]*deviceResources
+	deviceUsed  map[schedulingv1alpha1.DeviceType]*deviceResources
+	allocateSet map[string]*framework.PodInfo
+}
+
+// deviceResources is used to present resources per device.
+// we use the minor of device as key
+// "0": {koordinator.sh/gpu-core:100, koordinator.sh/gpu-memory-ratio:100, koordinator.sh/gpu-memory: 16GB}
+// "1": {koordinator.sh/gpu-core:100, koordinator.sh/gpu-memory-ratio:100, koordinator.sh/gpu-memory: 16GB}
+type deviceResources map[int]corev1.ResourceList
+
+func newNodeDeviceCache() *nodeDeviceCache {
+	return &nodeDeviceCache{
+		nodeDeviceInfos: make(map[string]*nodeDevice),
+	}
+}
+
+func newNodeDevice() *nodeDevice {
+	return &nodeDevice{
+		deviceTotal: make(map[schedulingv1alpha1.DeviceType]*deviceResources),
+		deviceFree:  make(map[schedulingv1alpha1.DeviceType]*deviceResources),
+		deviceUsed:  make(map[schedulingv1alpha1.DeviceType]*deviceResources),
+		allocateSet: make(map[string]*framework.PodInfo),
+	}
+}
+
+func (n *nodeDeviceCache) getNodeDevice(nodeName string) *nodeDevice {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	return n.nodeDeviceInfos[nodeName]
+}
+
+func (n *nodeDevice) resetDeviceFree(dType schedulingv1alpha1.DeviceType) {
+	for minor, usedResource := range *n.deviceUsed[dType] {
+		(*n.deviceFree[dType])[minor] = quotav1.SubtractWithNonNegativeResult(
+			(*n.deviceTotal[schedulingv1alpha1.GPU])[minor],
+			usedResource)
+	}
+}
+
+func (n *nodeDevice) updateDeviceUsed(dType schedulingv1alpha1.DeviceType, allocation *apiext.DeviceAllocation, add bool) {
+	if add {
+		(*n.deviceUsed[dType])[int(allocation.Minor)] = quotav1.Add(
+			(*n.deviceUsed[dType])[int(allocation.Minor)],
+			allocation.Resources)
+	} else {
+		(*n.deviceUsed[dType])[int(allocation.Minor)] = quotav1.SubtractWithNonNegativeResult(
+			(*n.deviceUsed[dType])[int(allocation.Minor)],
+			allocation.Resources)
+	}
+}
+
+func (n *nodeDevice) tryAllocateGPU(podRequest corev1.ResourceList) []*apiext.DeviceAllocation {
```

Comment:
tryAllocateGPU should be tryAllocateDevice not only GPU;  the return type should be "map[deviveType][]*apiext.DeviceAllocation",  expcept different device resource parse from podRequest, the assign logic is the same.  the process should be:  if a podRequest wants serveral devices,  only every device is satisfy can really lead to assign success.  you should automaticlly parse device resource from podRequest based on a configuration, for example, the config is {"gpu": ["gpu-core", "gpu-memory", "gpu-memory-ratio"],  "rdma": ["rdma"]},  so that you can parse different device resources from podRequest